src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,386 @@
+        fs::create_dir_all(&compact_dir_path).int_err()?;
+        Ok(compact_dir_path)
+    }
+
+    async fn replace_data_files(
+        &self,
+        dataset: Arc<dyn Dataset>,
+        chain_files_info: &ChainFilesInfo,
+    ) -> Result<(Vec<Url>, Multihash), CompactError> {
+        let chain = dataset.as_metadata_chain();
+
+        let mut new_head = chain_files_info.new_head.clone();
+        let mut old_data_slices: Vec<Url> = vec![];
+
+        for data_slice_batch_info in chain_files_info.data_slice_batches.iter().rev() {
+            let add_data_params = AddDataParams {
+                prev_checkpoint: None,
+                prev_offset: data_slice_batch_info.prev_offset,
+                new_offset_interval: data_slice_batch_info.new_offset_interval.clone(),
+                new_watermark: None,
+                new_source_state: None,
+            };
+
+            let commit_result = dataset
+                .commit_add_data(
+                    add_data_params,
+                    Some(OwnedFile::new(
+                        data_slice_batch_info.new_file_path.as_ref().unwrap(),
+                    )),
+                    None,
+                    CommitOpts {
+                        block_ref: &BlockRef::Head,
+                        system_time: Some(Utc::now()),
+                        prev_block_hash: Some(Some(&new_head)),
+                        check_object_refs: false,
+                        is_reset_head: false,
+                    },
+                )
+                .await
+                .int_err()?;
+            new_head = commit_result.new_head;
+
+            // Check is new commited block equals to old one
+            // It is possible when `max-slice-size` is smaller
+            // than size of one data slice
+            if data_slice_batch_info.data_slices_batch.len() == 1 {
+                let new_block = chain.get_block(&new_head).await.int_err()?;
+                if let MetadataEvent::AddData(add_event) = new_block.event {
+                    if let Some(data_slice) = add_event.new_data {
+                        if data_slice_batch_info
+                            .data_slices_batch
+                            .first()
+                            .unwrap()
+                            .to_string()
+                            .contains(&data_slice.physical_hash.as_multibase().to_string())
+                        {
+                            continue;
+                        }
+                    }
+                }
+            }
+            old_data_slices.extend(data_slice_batch_info.data_slices_batch.clone());
+        }
+
+        Ok((old_data_slices, new_head))
+    }
+
+    fn remove_old_files(
+        &self,
+        chain_files_info: &ChainFilesInfo,
+        old_data_slices: &Vec<Url>,
+    ) -> Result<(), CompactError> {
+        for file_url in old_data_slices {
+            fs::remove_file(file_url.to_file_path().unwrap()).int_err()?;
+        }
```

Comment:
1) Will not work for S3. S3/Minio URLs would not convert to file paths.
2) This isn't removing the checkpoint files. You are not even considering them in entire algorithm.

In general, is this step even nevessary? I understand we've left some garbage, but we are not pointing to it. We have plans to implement Garbage Collector in the datasets, maybe it would be enough?

Reply:
Agree, this is a tricky part that won't work correctly with remotes. @sergiimk @zaychenko-sergei What do you say if we currently remove this step and assign it as a separate issue??

Reply:
It is fine to skip it. We hope GC implementation will solve deleting all orphan files.